### PR TITLE
Hmusum/tunable function for confidence

### PR DESCRIPTION
**Pull request by hmusum**
* Initial values for the three new tunable variables makes the function give
  the same output as today's algorithm
* REST API for /controller/v1/jobs/upgrader now returns values for
  all 4 settings in Upgrader for GET requests

### CODE REVIEW

**Comment by bratseth** [comment]
Do we really need this in ZK or could we just use constants - observe on one release, modify the constants for the next.

Each of these things may be justifiable in isolation, but If we keep doing it most of our code will be managing dynamic knobs, and we won't be able to say much about the behavior of the system by looking at the code. And running it will be like

controlcenter

**Comment by mpolden** [comment]
@bratseth The problem is that we don't have enough data to decide on those constants. For the past few releases the current values appears to be have been fine.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/versions/VespaVersion.java [hunk 1]
```diff
@@ -136,4 +136,29 @@ public boolean equals(Object other) {
 
     }
 
+    private static boolean nonCanaryApplicationsBroken(ApplicationList failingOnThis,
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/versions/VespaVersion.java]: Consider adding a test that uses non-default values for the calculation.

**Comment by bratseth** [comment]
> The problem is that we don't have enough data to decide on those constants. 

Or that the data is too noisy, but this doesn't answer my point which is: Why isn't it good enough to be able to adjust the parameters for the *next* release?

**Comment by mpolden** [comment]
> Why isn't it good enough to be able to adjust the parameters for the next release?

Because otherwise someone must spend time "unbreaking" a release that has become broken by manually triggering upgrades until confidence is restored.

**Comment by bratseth** [comment]
> Because otherwise someone must spend time "unbreaking" a release that has become broken by manually triggering upgrades until confidence is restored.

Yes, but that problem will be gone in a month ... if not already.
I think that the releases we have seen after we started work on this shows that the problem is now much smaller than we thought, Harald fixing the overload problems, and the defect config server made things about as stable as in the historical data. So I suggest we don't need these knobs after all, and that we should also adjust the confidence threshold closer to what it is at this moment compared to what I suggested on Friday.

Or, if you really insist we could merge this and remove the dynamic part later.

**Comment by hmusum** [comment]
The adjustable parameters might be be helpful when we are unable to get out a new release (when we are stuck because we have too many model versions), especially when one or more of the older releases have bugs...  Or maybe we want a way of saying "ignore confidence, upgrade anyway" instead. 

BTW, confidence threshold in this PR is unchanged from what we have now, it's just implemented  another way and with the possibility of tuning it.

**Comment by bratseth** [comment]
> Or maybe we want a way of saying "ignore confidence, upgrade anyway" instead.

Yes, that would be simpler, and wouldn't let one bad situation screw up the convergence of these parameters ...

**Comment by mpolden** [comment]
> Or maybe we want a way of saying "ignore confidence, upgrade anyway" instead.

Seems good to me. Add API call for doing that and convert the current default values to constants?

**Comment by hmusum** [comment]
@mpolden please take a look

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/versions/VersionStatusTest.java [hunk 2]
```diff
@@ -235,6 +235,18 @@ public void testVersionConfidence() {
...
                      VespaVersion.Confidence.low, confidence(versions, Vtag.currentVersion));
+
+        // Same as above, but ignore confidence calculations, will force normal confidence
+        tester.controllerTester().curator().writeIgnoreConfidence(true);
+        versionStatus = VersionStatus.compute(tester.controller());
+        versions = versionStatus.versions();
+
+        assertEquals("No deployments: Low",
+                     VespaVersion.Confidence.low, confidence(versions, version0));
+        assertEquals("40% of defaults failed, but confidence is ignored: N  ormal",
```
Comment (mpolden) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/versions/VersionStatusTest.java]: "N  ormal" -> "Normal"

**Review by mpolden — COMMENTED** [review]
One minor comment. LGTM!

**Comment by bratseth** [comment]
Can we also set applicationsGivingMaxConfidence higher than 4?
That number means that when we have 4 applications (not 4 failing) we will have full confidence in our failure ratio being a true measure of the quality of the version and will use the failure ratio number unadjusted.

So, e.g if we have 4 applications and 1 of them fails we will say "broken".

I think you should set it to 40 to get a smoothed variant of the current behavior.
That would have worked with the last releases but from watching the last one closely I think it is a little too close for comfort, so I suggest 50.

**Comment by mpolden** [comment]
Close?

**Comment by hmusum** [comment]
Yeah, this didn't work out

**Pull request closed without merging**